Addition of Chromatograms class and chromData accessors

philouail opened the pull request:
So here are the general filtering functions. 

They only allow to filter through numerical values but I thought about it and using just base R code is enough to filter for  character values I believe. 

Tell me what you think.

jorainer commented on R/ChromBackend.R:
```diff
@@ -219,6 +241,16 @@
 #'   are not supported for `dataStorage()`.
 #'
+#' - `filterChromData()`: filters any numerical chromatographic data variables
+#'   the based on the provided numerical `ranges`. The method should return a
```
there is one *the* too much.

jorainer commented on R/ChromBackend.R:
```diff
@@ -219,6 +241,16 @@
+#' - `filterPeaksData()`: filters any numerical peaks data variables based on
+#'   the provided numerical `ranges`. The method should return a `ChromBackend`
+#'   object with the chromatograms that match the condition. This function will
+#'   results in an object with the same number of chromatogram as the original.
```
Maybe add *... as the original, but eventually less retention time - intensity value pairs.*. Otherwise it might not be clear what is actually filtered.

jorainer commented on R/ChromBackend.R:
```diff
@@ -796,5 +828,87 @@ setMethod("split", "ChromBackend", function(x, f, drop = FALSE, ...) {
+#' Filter chromData removes chromatograms that do not match the filter criteria.
+#' FilterPeaks Data does not, only some data points are removed.
+#'
+#' can i make it possible that it can filter both character and numeric columns?
```
agree. it makes more sense with numerical values - especially for the `filterChromData()` it would be easy to do that with *standard* R conditions. The advantage of the `filterChromData()` function is that we can maybe have optimized versions for specific backends. Example: a SQL backend might implement the filter function such that the filter is done as an SQL statement, so the database takes care of filtering, and no data needs to be loaded/processed in R.

jorainer commented on R/ChromBackend.R:
```diff
@@ -796,5 +828,87 @@ setMethod("split", "ChromBackend", function(x, f, drop = FALSE, ...) {
+#' @rdname ChromBackend
+#'
+#' @export
+setMethod("filterPeaksData", "ChromBackend",
```
the function is great. My only concern: do we want to apply this filter immediately or do we want to do it similar to `Spectra`, i.e. using a processing queue in the `Chromatograms` object to cache this operation?

The main reason we might want that is for *read only* backends, i.e. such as SQL database(s), backends reading on-the-fly from mzML or similar files.

jorainer replied:
I think we could have this method only implemented for `Chromatograms()` actually, not for the backend. All the checking for input parameters could be done there and then we simply add the `.filter_ranges()` function to the processing queue. Or we could have a parameter `lazy = FALSE` (or maybe better `immediate = TRUE`?) that would bypass the processing queue and apply the filter immediately? Let's discuss in the next meeting.

philouail replied:
yesss I think that makes sense.. Some chromatograms object are going to be big. I think it's good to prepare for that already.

jorainer reviewed (commented):
Looks good, but I would like to discuss whether we should implement that for the backend or for `Chromatograms`.

philouail commented:
@jorainer So I have removed and added some things. This PR now is comprised of:

- filtering of `chromData` on the backend 
- addition of `[[` and `[[<-` method
-  addition of `Chromatograms` class and function to create the object 
- addition of `chromData` accessor and filtering on a `Chromatograms` object. 

It is lacking information regarding peaksData as thisis a bigger implementation and will be for next PR. 

Lastly, I'm uncertain about the documentation of the `Chromatograms` and `Chromatograms -chromData`. how to link everything together. I also think that `chromData` and `peaksData` are nicely explained in the `chromBackend` documentation, is it really necessary to also have it in the `Chromatograms` documentation ?

jorainer commented:
Regarding documentation and whether to have duplicated documentation in `ChromBackend` and `Chromatograms`: that's more a general discussion. The user will use mostly (or exclusively) `Chromatograms`, so IMHO all methods should be documented there to make it easier for the user to find everything in one place. The documentation in `ChromBackend` are more for the developers to guide them for new implementations. I would suggest to have documentation, e.g. for `chromData()` in both - for `Chromatograms` more tailored to the user, in `ChromBackend` more for the developer. There can be duplication I think.

jorainer commented on R/Chromatograms-chromData.R:
```diff
@@ -0,0 +1,274 @@
+#'
+#' @name chromData
+#'
+#' @aliases chromData
```
If you have here the `chromData` alias you should (I think) remove it from the *AllGenerics.R*, i.e. don't add the hidden_aliases tag there. Same for all other generic methods. We should avoid the issues we have/had in *Spectra* right from the beginning, i.e. where `?spectraData` would open up the hidden aliases documentation, not the one for `Spectra`.

philouail replied:
It's now removed and fixed ! I have left it for `reset`, it is not really implemented (only in ChromBackend) so I want to see how user should use it before I document it.

jorainer commented on R/Chromatograms-chromData.R:
```diff
@@ -0,0 +1,274 @@
+#' @include Chromatograms.R hidden_aliases.R
+
+#' @title Accessing chromatographic variables.
```
I like the idea to split the `Chromatograms` documentation into *chrom variables* and *peak variables*. That's different to *Spectra*  where we have all in the `Spectra` documentation (which then became very large). But then maybe reformulate the title into something like `@title Chromatographic Data` and ideally also link to this documentation directly in the `Chromatrograms` documentation (i.e. in the general section mention that more information on chromatographic variables can be found in the `chromData()` documentation and information on peaks variables in `peaksData()`.

philouail replied:
What do you think of `Chromatographic Metadata` instead ? I fear just saying data is a bit confusing as we do not access RT and intensity using `chromData()`. Any thought ?

philouail replied:
I just want to find a clear way to separate peaksData and chromData. 

In my head Chromatographic Data = peaksData + chromData

jorainer replied:
ah, yes, good point. Yes, chromatographic metadata makes sense.

jorainer commented on vignettes/using-a-chromatograms-object.Rmd:
```diff
@@ -0,0 +1,54 @@
+be
+
+#' create Chromatograms object 
+chrom <- Chromatogram(be)
```
here it should be `Chromatograms(be)`, no?

philouail replied:
i removed the vignette for now ! Will prepare it for after `peaksData` is implemented

jorainer commented on R/AllGenerics.R:
```diff
@@ -1,48 +1,41 @@
-#' @rdname ChromBackend
 setGeneric("chromData", function(object, ...) standardGeneric("chromData"))
-#' @rdname ChromBackend
+#' @rdname hidden_aliases
```
Maybe we can have aliases for all these (chrom data accessor/setter) methods directly in the documentation for `chromData`? Then e.g. `?chromIndex` would redirect the user directly there - and the documentation describing this function should be provided there. Otherwise users would end up in the "hidden_aliases" man page - just like what happened in Spectra...

philouail replied:
So i only have hidden_aliases tags for the replacement function, for the others, it links directly to the right documentation. The replacement functions behaves a bit weird which is why i had the hidden_alias tag. But i'll try to look more into it and fix it.

jorainer reviewed (changes requested):
Really nice PR, thanks Phili!

I have some minor comments/suggestions, mostly related to the documentation and how to organize that.

jorainer reviewed (approved):
Yey! looks good! thanks!
